src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/CSharpVirtualDocument.cs

```diff
@@ -82,6 +90,42 @@ public override VirtualDocumentSnapshot Update(IReadOnlyList<TextChange> changes
             return _currentSnapshot;
         }
 
-        private CSharpVirtualDocumentSnapshot UpdateSnapshot() => new CSharpVirtualDocumentSnapshot(Uri, TextBuffer.CurrentSnapshot, HostDocumentSyncVersion);
+        private bool TryRevertProvisionalChanges()
+        {
+            if (!_hasProvisionalChanges)
+            {
+                return false;
+            }
+
+            Debug.Assert(_previousSnapshot != null);
+
+            using var revertEdit = TextBuffer.CreateEdit(EditOptions.None, _previousSnapshot.Snapshot.Version.VersionNumber, InviolableEditTag.Instance);
+            var previousChanges = _previousSnapshot.Snapshot.Version.Changes;
+            for (var i = 0; i < previousChanges.Count; i++)
+            {
+                var change = previousChanges[i];
+                revertEdit.Replace(change.NewSpan, change.OldText);
```

Comment:
So these changes are always expected to be replaces?

Reply:
This takes care of all types of edits. Insert and Delete are basically Replaces.